docs/en/sql-reference/functions/conditional-functions.md

```diff
@@ -267,3 +267,177 @@ SELECT clamp(1, 2, 3) result,  toTypeName(result) type;
+
+This form allows full flexibility and is internally implemented using the [multiIf](#multiif-multiif) function. Each condition is evaluated independently, and expressions can include non-constant values.
+
+```sql
+SELECT
+    number,
+    CASE
+        WHEN number % 2 = 0 THEN number + 1
+        WHEN number % 2 = 1 THEN number * 10
+        ELSE number
+    END AS result
+FROM system.numbers
+WHERE number < 5;
+
+-- is translated to
+SELECT
+    number,
+    multiIf((number % 2) = 0, number + 1, (number % 2) = 1, number * 10, number) AS result
+FROM system.numbers
+WHERE number < 5
+
+┌─number─┬─result─┐
+│      0 │      1 │
+│      1 │     10 │
+│      2 │      3 │
+│      3 │     30 │
+│      4 │      5 │
+└────────┴────────┘
+
+5 rows in set. Elapsed: 0.002 sec.
+```
+
+2. `CASE <expr> WHEN <val1> THEN ... WHEN <val2> THEN ... ELSE ... END`
+
+This more compact form is optimized for constant value matching and internally uses `caseWithExpression()`. However, it has limitations:
+
+- All return expressions must be constants or deterministic expressions not depending on the input.
+- Using expressions like number + 1 or timezone-aware functions will result in an error.
+
+For example, the following is valid:
+
+```sql
+SELECT
+    number,
+    CASE number
+        WHEN 0 THEN 100
+        WHEN 1 THEN 200
+        ELSE 0
+    END AS result
+FROM system.numbers
+WHERE number < 3;
+
+-- is translated to
+
+SELECT
+    number,
+    caseWithExpression(number, 0, 100, 1, 200, 0) AS result
+FROM system.numbers
+WHERE number < 3
+
+┌─number─┬─result─┐
+│      0 │    100 │
+│      1 │    200 │
+│      2 │      0 │
+└────────┴────────┘
+
+3 rows in set. Elapsed: 0.002 sec.
+```
+
+Conversely, the following is invalid:
+
+```sql
+SELECT
+    number,
+    CASE number
+        WHEN 0 THEN number + 1
+        WHEN 1 THEN number * 10
+        ELSE number
+    END
+FROM system.numbers
+WHERE number < 3;
+
+-- is translated to
+
+SELECT
+    number,
+    caseWithExpression(number, 0, number + 1, 1, number * 10, number)
+FROM system.numbers
+WHERE number < 3
+
+Elapsed: 0.001 sec.
+
+Received exception:
+Code: 44. DB::Exception: Argument at index 2 for function transform must be constant. (ILLEGAL_COLUMN)
+
+```
+
+### Caveats  {#caveats}
+
+ClickHouse determines the result type of a CASE expression (or its internal equivalent, such as `multiIf`) before evaluating any conditions. This is important when the return expressions differ in type, such as different timezones or numeric types.
+
+- The result type is selected based on the largest compatible type among all branches.
+- Once this type is selected, all other branches are implicitly cast to it—even if their logic would never be executed at runtime.
+- For types like DateTime64, where the timezone is part of the type signature, this can lead to surprising behavior: the first encountered timezone may be used for all branches, even when other branches specify different timezones.
```

Comment:
```suggestion
- For types like [DateTime64](/sql-reference/data-types/datetime64), where the timezone is part of the type signature, this can lead to surprising behavior: the first encountered timezone may be used for all branches, even when other branches specify different timezones.
```